File: python/lisk32.py
# Copyright (c) 2021 hirish
# Copyright (c) 2017, 2020 Pieter Wuille
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""Reference implementation for Lisk32 addresses. (based on bech32)"""


from enum import Enum

class Encoding(Enum):
    """Enumeration type to list the various supported encodings."""
    BECH32 = 1

CHARSET = "zxvcpmbn3465o978uyrtkqew2adsjhfg"
HRP = 'lsk'

def bech32_polymod(values):
    """Internal function that computes the Bech32 checksum."""
    generator = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3]
    chk = 1
    for value in values:
        top = chk >> 25
        chk = (chk & 0x1ffffff) << 5 ^ value
        for i in range(5):
            chk ^= generator[i] if ((top >> i) & 1) else 0
    return chk

def bech32_verify_checksum(data):
    """Verify a checksum given converted data characters."""
    const = bech32_polymod(data)
    if const == 1:
        return Encoding.BECH32
    return None

def bech32_create_checksum(data):
    """Compute the checksum values of given data."""
    const = 1
    polymod = bech32_polymod(data + [0, 0, 0, 0, 0, 0]) ^ const
    return [(polymod >> 5 * (5 - i)) & 31 for i in range(6)]


def bech32_encode(hrp, data):
    """Compute a Lisk32 string given HRP and data values."""
    combined = data + bech32_create_checksum(data)
    return hrp + ''.join([CHARSET[d] for d in combined])

def bech32_decode(bech):
    """Validate a Lisk32 string, and determine HRP and data."""
    if ((any(ord(x) < 33 or ord(x) > 126 for x in bech)) or
            (bech.lower() != bech and bech.upper() != bech)):
        return (None, None, None)
    bech = bech.lower()
    pos = 3
    if not all(x in CHARSET for x in bech[pos:]):
        return (None, None, None)
    hrp = bech[:pos]
    data = [CHARSET.find(x) for x in bech[pos:]]
    spec = bech32_verify_checksum(data)
    if spec is None:
        return (None, None, None)
    return (hrp, data[:-6], spec)

def convertbits(data, frombits, tobits, pad=False):
    """General power-of-2 base conversion."""
    acc = 0
    bits = 0
    ret = []
    maxv = (1 << tobits) - 1
    max_acc = (1 << (frombits + tobits - 1)) - 1
    for value in data:
        if value < 0 or (value >> frombits):
            return None
        acc = ((acc << frombits) | value) & max_acc
        bits += frombits
        while bits >= tobits:
            bits -= tobits
            ret.append((acc >> bits) & maxv)
    if pad:
        if bits:
            ret.append((acc << (tobits - bits)) & maxv)
    elif bits >= frombits or ((acc << (tobits - bits)) & maxv):
        return None
    return ret


def decode(addr):
    """Decode a Lisk32 address."""
    hrpgot, data, spec = bech32_decode(addr)
    if hrpgot != HRP:
        return None
    decoded = convertbits(data, 5, 8, False)
    if decoded is None or len(decoded) != 20:
        return None
    if spec != Encoding.BECH32:
        return None
    return decoded


def encode(addr):
    """Encode a Lisk32 address."""
    ret = bech32_encode(HRP, convertbits(addr, 8, 5))
    if decode(ret) == None:
        return None
    return ret

File: csharp/Lisk32.cs
/* Copyright (c) 2017 Pieter Wuille
 * Copyright (c) 2021 hirishh
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Security.Cryptography;

namespace Lisk32
{
	class Lisk32ConversionException: Exception
	{
		public Lisk32ConversionException()
			: base() {
		}
		public Lisk32ConversionException(String message)
			: base(message) {
		}
	}
	public static class Converter
	{
		private const string HRT = "lsk";
		private const string CHARSET_BECH32 = "zxvcpmbn3465o978uyrtkqew2adsjhfg";
		// https://go.dev/play/p/SsuX334se98
		private static readonly sbyte[] DICT_BECH32 = new sbyte[128]{
			-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
			-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
			-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
			-1, -1, 24,  8,  9, 11, 10, 14, 15, 13, -1, -1, -1, -1, -1, -1,
			-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
			-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
			-1, 25,  6,  3, 26, 22, 30, 31, 29, -1, 28, 20, -1,  5,  7, 12,
			 4, 21, 18, 27, 19, 16,  2, 23,  1, 17,  0, -1, -1, -1, -1, -1,
		};
		private static uint PolyMod(byte[] input) {
			uint startValue = 1;
			for (uint i = 0; i < input.Length; i++) {
				uint c0 = startValue >> 25;
				startValue = (uint)(((startValue & 0x1ffffff) << 5) ^
					(input[i]) ^
					(-((c0 >> 0) & 1) & 0x3b6a57b2) ^
					(-((c0 >> 1) & 1) & 0x26508e6d) ^
					(-((c0 >> 2) & 1) & 0x1ea119fa) ^
					(-((c0 >> 3) & 1) & 0x3d4233dd) ^
					(-((c0 >> 4) & 1) & 0x2a1462b3));
			}
			return startValue ^ 1;
		}

		public static byte[] Combine(byte[] first, byte[] second)
		{
			byte[] ret = new byte[first.Length + second.Length];
			Buffer.BlockCopy(first, 0, ret, 0, first.Length);
			Buffer.BlockCopy(second, 0, ret, first.Length, second.Length);
			return ret;
		}
		private static byte[] createChecksum(byte[] data) {
			uint mod = PolyMod(data);
			byte[] ret = new byte[6];
			for (int i = 0; i < 6; i++) {
				ret[i] = (byte) ((mod >> (5 * (5 - i))) & 31);
			}
			return ret;
		}
		private static bool verifyChecksum(byte[] dataWithChecksum) {
			return PolyMod(dataWithChecksum) == 0;
		}
		private static int convertBits(byte[] bytes, int inBits, uint outBits, byte[] converted, int len, bool pad=true) {
			//byte[] converted = new byte[bytes.Length * inBits / outBits + (bytes.Length * inBits % outBits != 0 ? 1 : 0)];
			uint bits = 0, maxv = (uint) ((1 << (int) outBits) - 1), val = 0, inPos = 0, outPos=0;
			//int len = bytes.Length - inPos;
			while (len-- != 0) {
				val = (val << inBits) | bytes[inPos++];
				bits += (uint) inBits;
				while (bits >= outBits) {
					bits -= outBits;
					converted[outPos++] = (byte) ((val >> (int) bits) & maxv);
				}
			}
			if (pad) {
				if (bits != 0) {
					converted[outPos++] = (byte)((val << (int)(outBits - bits)) & maxv);
				}
			} else if ((((val << (int) (outBits - bits)) & maxv) != 0) || bits >= inBits) {
				throw new Lisk32ConversionException("Bit conversion error! bits:" + bits + " | inBits:" + inBits + " - " + ((val << (int) (outBits - bits)) & maxv));
			}
			return (int) outPos;
		}

		// address = 20 byte Lisk address
		public static string EncodeLisk32(byte[] address)
		{
			if (address.Length != 20) {
				throw new Lisk32ConversionException("Invalid lisk address!");
			}
			// byte[] data = new byte[100];
			byte[] data = Enumerable.Repeat((byte)0x00, 100).ToArray();
			int len = convertBits(address, 8, 5, data, address.Length);
			Array.Resize(ref data, len + 6); // Add checksum zeros
			byte[] checksum = createChecksum(data);

			StringBuilder ret = new StringBuilder(HRT);
			for (int i = 0; i < len; i++) {
				ret.Append(CHARSET_BECH32[data[i]]);
			}
			for (int i = 0; i < 6; i++) {
				ret.Append(CHARSET_BECH32[checksum[i]]);
			}
			return ret.ToString();
		}

		public static byte[] DecodeLisk32(string addr) {
			if (!addr.StartsWith(HRT)) {
				throw new Lisk32ConversionException("Invalid Lisk32 address: wrong prefix (lsk)");
			}
			int dataLen = addr.Length - 3;
			byte[] data = new byte[dataLen];
			for (int i = 0; i < dataLen; i++) {
				data[i] = (byte) addr[3 + i];
			}
			sbyte err = 0;
			for (int i = 0; i < dataLen; i++) {
				sbyte k = DICT_BECH32[data[i]];
				err |= k;
				data[i] = unchecked((byte) k);
			}
			if (err == -1 || !verifyChecksum(data)) {
				throw new Lisk32ConversionException("Invalid Lisk32 address: wrong checksum");
			}
			byte[] decoded = new byte[40];
			int decodedLen = convertBits(data, 5, 8, decoded, dataLen - 6);
			if (decodedLen != 20) {
				throw new Lisk32ConversionException("Invalid Lisk32 address: wrong decoded length");
			}
			byte[] final = new byte[decodedLen];
			System.Buffer.BlockCopy(decoded, 0, final, 0, decodedLen);
			return final;
		}
	}
}
